rift/node.py

```diff
@@ -2378,51 +2495,68 @@ def defer_spf_timer_expired(self):
         end_tie_id = packet_common.make_tie_id(direction, system_id, prefix_type,
                                                packet_common.MAX_U32)
+
         node_tie_ids = self.tie_packet_infos.irange(start_tie_id, end_tie_id, (True, True))
+
         for node_tie_id in node_tie_ids:
             node_tie_packet_info = self.tie_packet_infos[node_tie_id]
             node_tie_packet = node_tie_packet_info.protocol_packet.content.tie
             node_ties.append(node_tie_packet)
+
         return node_ties
 
     def node_ties(self, direction, system_id):
         # Return an ordered list of all node TIEs from the given node and in the given direction
         return self.ties_of_type(direction, system_id, common.ttypes.TIETypeType.NodeTIEType)
 
-    def node_neighbors(self, node_ties, neighbor_direction):
+    def node_neighbors(self, node_ties, neighbor_directions):
         # A generator that yields (nbr_system_id, nbr_tie_element) tuples for all neighbors in the
-        # specified direction of the nodes in the node_ties list.
+        # specified directions of the nodes in the node_ties list.
         for node_tie in node_ties:
             node_level = node_tie.element.node.level
+
             for nbr_system_id, nbr_tie_element in node_tie.element.node.neighbors.items():
                 nbr_level = nbr_tie_element.level
-                if neighbor_direction == constants.DIR_SOUTH:
+
+                correct_direction = False
+
+                if constants.DIR_SOUTH in neighbor_directions:
                     correct_direction = (nbr_level < node_level)
-                elif neighbor_direction == constants.DIR_NORTH:
+
+                if not correct_direction and constants.DIR_NORTH in neighbor_directions:
                     correct_direction = (nbr_level > node_level)
-                elif neighbor_direction == constants.DIR_EAST_WEST:
+
+                if not correct_direction and constants.DIR_EAST_WEST in neighbor_directions:
                     correct_direction = (nbr_level == node_level)
-                else:
-                    assert False
+
                 if correct_direction:
-                    yield (nbr_system_id, nbr_tie_element)
+                    yield nbr_system_id, nbr_tie_element
 
     def spf_run(self):
+
         self._spf_runs_count += 1
         # TODO: Currently we simply always run both North-SPF and South-SPF, but maybe we can be
         # more intelligent about selectively triggering North-SPF and South-SPF separately.
-        self.spf_run_direction(constants.DIR_SOUTH)
-        self.spf_run_direction(constants.DIR_NORTH)
+
+        # Run normal SPFs (south and north)
+        self.spf_run_direction(constants.DIR_SOUTH, special_for_neg_disagg=False)
+        self.spf_run_direction(constants.DIR_NORTH, special_for_neg_disagg=False)
+
+        # Run special SPF (southbound) for negative disaggregation only if current node is a ToF
+        if self.top_of_fabric():
```

Comment:
@brunorijsman is this the right way to check if a node is a ToF? (to run special SPF)

Reply:
Yes, calling top_of_fabric() is the right thing to do, but note that there is a TODO in the implementation of top_of_fabric (file neighbor.py)

    def top_of_fabric(self):
        # TODO: Is this right? Should we look at capabilities.hierarchy_indications?
        return self.level == common.constants.top_of_fabric_level